nchelpers/__init__.py

```diff
@@ -331,56 +406,117 @@ def climo_periods(self):
                 }
 
     @property
+    def ensemble_member(self):
+        """CMIP5 standard ensemble member code for this file"""
+        template = 'r{r}i{i}p{p}'
+        if self.is_unprocessed_gcm_output:
+            return template.format(r=self.realization,
+                                   i=self.initialization_method,
+                                   p=self.physics_version)
+        else:
+            return template.format(r=self.driving_realization,
+                                   i=self.driving_initialization_method,
+                                   p=self.driving_physics_version)
+
+    def _cmor_type_filename_components(self, tres_to_mip_table=standard_tres_to_mip_table, **override):
+        """Return a dict containing appropriate arguments to function cmor_type_filename (q.v.),
+        with content built from this file's metadata.
+
+        :param tres_to_mip_table: (dict) a dict mapping time resolution (as computed by the property
+            self.time_resolution) to a valid MIP table name.
+        :param override: keyword arguments that can override or extend the base components computed here.
+        :return: (dict) as above
+        """
+
+        # File content-independent components
+        components = {
+            'variable': '+'.join(sorted(self.dependent_varnames)),
+            'ensemble_member': self.ensemble_member,
+        }
+
+        # Components depending on the type of file
+        if self.is_multi_year_mean:
+            components.update(
+                time_range=_cmor_formatted_time_range(dateutil.parser.parse(self.climo_start_time),
+                                                      dateutil.parser.parse(self.climo_end_time)),
+                frequency=self.frequency
+            )
+        else:
+            # Regarding how the 'mip_table' component is defined here, see the discussion in section titled
+            # "MIP table / table_id" in
+            # https://pcic.uvic.ca/confluence/display/CSG/PCIC+metadata+standard+for+downscaled+data+and+hydrology+modelling+data
+            # Specifically, we do not consult the value of the attribute table_id because it is too limited for our
+            # needs. Instead we map the file's time resolution to a value.
+            components.update(
+                time_range=self.time_range_formatted,
+                mip_table = tres_to_mip_table and tres_to_mip_table.get(self.time_resolution, None)
+            )
+
+        if self.is_unprocessed_gcm_output:
```

Comment:
So, I don't normally advocate for an OO design, but a lot of things in the file seem to hint that a subclassing design would be appropriate. We have a variety of different object types that are determined by the ``is_`` methods (``is_downscaling_output()``, ``is_unprocessed_gcm_output()``, etc.) and then a bunch of other methods that change their behaviour based on what time of object it is. That's classic OO behaviour and would be handled much more elegantly with a few simple subclasses that have their own implementations of ``_cmor_type_filename_components()`` and ``ensemble_member()``. With only *2* methods that would be in the subclasses, it may not be worth it, but if you find yourself engaging in this pattern with any more methods, you may want to give it some consideration.